rmw_opensplice_cpp/src/rmw_waitset.cpp

```diff
@@ -0,0 +1,209 @@
+    static_cast<DDS::ConditionSeq *>(rmw_allocate(sizeof(DDS::ConditionSeq)));
+  if (!waitset_info->active_conditions) {
+    RMW_SET_ERROR_MSG("failed to allocate active_conditions sequence");
+    goto fail;
+  }
+
+  waitset_info->attached_conditions =
+    static_cast<DDS::ConditionSeq *>(rmw_allocate(sizeof(DDS::ConditionSeq)));
+  if (!waitset_info->attached_conditions) {
+    RMW_SET_ERROR_MSG("failed to allocate attached_conditions sequence");
+    goto fail;
+  }
+
+  // If max_conditions is greater than zero, re-allocate both ConditionSeqs to max_conditions
+  if (max_conditions > 0) {
+    RMW_TRY_PLACEMENT_NEW(
+      waitset_info->active_conditions, waitset_info->active_conditions, goto fail,
+      DDS::ConditionSeq, max_conditions)
+
+    RMW_TRY_PLACEMENT_NEW(
+      waitset_info->attached_conditions, waitset_info->attached_conditions, goto fail,
+      DDS::ConditionSeq,
+      max_conditions)
+  } else {
+    // Else, don't preallocate: the vectors will size dynamically when rmw_wait is called.
+    // Default-construct the ConditionSeqs.
+    RMW_TRY_PLACEMENT_NEW(
+      waitset_info->active_conditions, waitset_info->active_conditions, goto fail,
+      DDS::ConditionSeq)
+
+    RMW_TRY_PLACEMENT_NEW(
+      waitset_info->attached_conditions, waitset_info->attached_conditions, goto fail,
+      DDS::ConditionSeq)
+  }
+
+  waitset->fixed_guard_conditions = fixed_guard_conditions;
+  if (fixed_guard_conditions) {
+    if (!fixed_guard_conditions->guard_conditions) {
+      RMW_SET_ERROR_MSG("Received invalid guard condition array");
+      goto fail;
+    }
+    // We also need to attach the fixed guard conditions to the waitset (and detach them in destroy)
+    for (size_t i = 0; i < fixed_guard_conditions->guard_condition_count; ++i) {
+      void * guard_cond = fixed_guard_conditions->guard_conditions[i];
+      if (!guard_cond) {
+        RMW_SET_ERROR_MSG("Received invalid guard condition");
+        goto fail;
+      }
+      dds_guard_cond = static_cast<DDS::GuardCondition *>(guard_cond);
+      ret = waitset_info->waitset->attach_condition(dds_guard_cond);
+      if (ret != DDS::RETCODE_OK) {
+        RMW_SET_ERROR_MSG("Failed to attach condition to waitset");
+      }
+    }
+  }
+
+  return waitset;
+
+fail:
+  if (waitset_info) {
+    if (waitset_info->active_conditions) {
+      // How to know which constructor threw?
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        waitset_info->active_conditions->~DDS_DCPSUObjSeq(), ConditionSeq)
+      rmw_free(waitset_info->active_conditions);
+    }
+    if (waitset_info->attached_conditions) {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        waitset_info->attached_conditions->~DDS_DCPSUObjSeq(), ConditionSeq)
+      rmw_free(waitset_info->attached_conditions);
+    }
+    if (waitset_info->waitset) {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(waitset_info->waitset->~WaitSet(), WaitSet)
+      rmw_free(waitset_info->waitset);
+    }
```

Comment:
Oh, I didn't interpret your original comment correctly. Should I set all freed member variables to zero/null?